gr-digital/lib/linear_equalizer_impl.cc

```diff
@@ -0,0 +1,231 @@
+                  input_samples + num_inputs,
+                  in_prepended_history.begin() + d_num_taps - 1);
+        samples = in_prepended_history.data();
+    }
+
+    unsigned int tag_index = 0;
+
+    unsigned int j = 0;
+    size_t k, l = d_taps.size();
+    for (unsigned i = 0; i < nout; i++) {
+        output_symbols[i] = filter(&samples[j]);
+
+        if (taps) {
+            std::copy(d_taps.begin(), d_taps.end(), taps + d_num_taps * i);
+        }
+
+        if (state) {
+            state[i] = (unsigned short)d_training_state;
+        }
+
+        if (training_start_samples.size() > 0 &&
+            tag_index < training_start_samples.size()) {
+            unsigned int tag_sample = training_start_samples[tag_index];
+            if (tag_sample >= j && tag_sample < (j + decimation())) {
+                d_training_state = equalizer_state_t::TRAINING;
+                d_training_sample = 0;
+                tag_index++;
+            }
+        }
+
+        // Are we done with the training sequence
+        if (d_training_sample >= d_training_sequence.size()) {
+            if (d_adapt_after_training) {
+                d_training_state = equalizer_state_t::DD;
+            } else {
+                d_training_state = equalizer_state_t::IDLE;
+            }
+            d_training_sample = -1;
+        }
+
+        // Adjust taps
+        if (d_training_state == equalizer_state_t::TRAINING) {
+            d_decision = d_training_sequence[d_training_sample];
+            d_error = d_alg->error_tr(output_symbols[i],
+                                      d_training_sequence[d_training_sample++]);
+        } else if (d_training_state == equalizer_state_t::DD) {
+            d_error = d_alg->error_dd(output_symbols[i], d_decision);
+        }
+        switch (d_training_state) {
+        case equalizer_state_t::IDLE:
+            d_error = gr_complex(0.0, 0.0);
+            break;
+        case equalizer_state_t::TRAINING:
+        case equalizer_state_t::DD:
+            d_alg->update_taps(&d_taps[0], &samples[j], d_error, d_decision, l);
+            for (k = 0; k < l; k++) {
+                // Update aligned taps in filter object.
+                filter::kernel::fir_filter_ccc::update_tap(d_taps[k], k);
+            }
+            break;
+        }
+
+        j += decimation();
+    }
+
+    return nout;
+}
+
+int linear_equalizer_impl::work(int noutput_items,
+                                gr_vector_const_void_star& input_items,
+                                gr_vector_void_star& output_items)
+{
+    auto in = static_cast<const gr_complex*>(input_items[0]);
+    auto out = static_cast<gr_complex*>(output_items[0]);
+
+    int outlen = output_items.size();
+
+    unsigned short* state = nullptr;
+    gr_complex* taps = nullptr;
+
+    if (outlen > 1)
+        taps = static_cast<gr_complex*>(output_items[1]);
+    if (outlen > 2)
+        state = static_cast<unsigned short*>(output_items[2]);
+
+    if (d_updated) {
```

Comment:
Again: undefined behaviour.

In practice probably fine, but many hairs have been pulled in the past from "surely my arch will do the right thing here?".